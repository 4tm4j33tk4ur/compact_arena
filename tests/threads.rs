extern crate crossbeam_utils;

use compact_arena::mk_nano_arena;
use crossbeam_utils::thread::scope;

// With crossbeam's `scope`d threads, it is even possible to share an arena
// and its indices between multiple threads.
#[test]
fn test_scoped_arena() {
    mk_nano_arena!(arena);
    let i = arena.add(1usize);
    let v = scope(|s| {
        // spawn a thread using both the arena and an existing index
        let t = s.spawn(|_| {
            let u = arena[i];
            arena.add(u + 1)
        });
        // join it, receiving a new index from the thread
        t.join().unwrap()
    })
    .unwrap();
    assert_eq!(2, arena[v]);
}
