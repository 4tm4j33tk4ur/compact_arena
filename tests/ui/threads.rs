extern crate compact_arena;

use std::{thread, sync::mpsc::{channel, Receiver, Sender}};
use compact_arena::{mk_tiny_arena, Idx16};

fn mixup(s: Option<Sender<Idx16<'_>>>, r: Option<Receiver<Idx16<'_>>>) {
    mk_tiny_arena!(arena);
    match (s, r) {
        (Some(s), None) => s.send(arena.add(1usize)).unwrap(),
        (None, Some(r)) => assert_eq!(1, arena[r.recv().unwrap()]),
        _ => unreachable!()
    }
}

fn main() {
    let (s, r) = channel();
    let st = thread::spawn(|| mixup(Some(s), None));
    let rt = thread::spawn(|| mixup(None, Some(r)));
    st.join().unwrap();
    rt.join().unwrap();
}
