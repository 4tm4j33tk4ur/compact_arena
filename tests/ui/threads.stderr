error[E0623]: lifetime mismatch
 --> $DIR/threads.rs:9:35
  |
6 | fn mixup(s: Option<Sender<Idx16<'_>>>, r: Option<Receiver<Idx16<'_>>>) {
  |                           ---------                       --------- these two types are declared with different lifetimes...
...
9 |         (Some(s), None) => s.send(arena.add(1usize)).unwrap(),
  |                                   ^^^^^^^^^^^^^^^^^ ...but data from `r` flows into `s` here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0623`.
